### PR TITLE
Task2 javascript

### PR DESCRIPTION
**Pull request by denis1222223**

### CODE REVIEW

File: Task2-Javascript/src/first.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+function findFirst(arr, condition) {
+	for (var i = 0; i < arr.length; i++) {
```
Comment (ilyahumbar) [inline Task2-Javascript/src/first.js]: Use 'fold' to solve this problem. 
Rename file to correspond class/function name

File: Task2-Javascript/src/lazy-evaluation.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+function lazyEvaluation (func) {
+  return func.bind.apply(func, arguments);
```
Comment (ilyahumbar) [inline Task2-Javascript/src/lazy-evaluation.js]: Once evaluated, lazy should not run given function once again. It should store the result of evaluation and return the same result without calling target function (which might be expensive)
Consider using already implemented methods to solve this task (e.g. partial application)

File: Task2-Javascript/tests/memoization.Test.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    
+	it('lazy memo evaluation of sum 7 and 8', function () {
+        expect(lazyMemoSumSevenAndEight()).toBe(15);
+    });
+	
+	it('lazy memo evaluation of sum 2 and 3', function () {
+        expect(lazyMemoMultTwoAndThree()).toBe(6);
+    });
+	
+});
```
Comment (ilyahumbar) [inline Task2-Javascript/tests/memoization.Test.js]: Tests doesn't cover one important aspect. Target function should be called only once.
Also test scenarios when function returns falsy value (0, false, null, undefined). These are valid results of anu function call and they are should be memoized  as well. (Seems your code doesn't cover these cases)

File: Task2-Javascript/tests/lazy-evaluation.Test.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    
+	it('lazy evaluation of sum 7 and 8', function () {
+        expect(lazySumSevenAndEight()).toBe(15);
+    });
+	
+	it('lazy evaluation of sum 2 and 3', function () {
+        expect(lazyMultTwoAndThree()).toBe(6);
+    });
+	
+});
```
Comment (ilyahumbar) [inline Task2-Javascript/tests/lazy-evaluation.Test.js]: Test that function is called only once.
e.g. after calling 
lazyMultTwoAndThree()
lazyMultTwoAndThree()
lazyMultTwoAndThree()

function sum should be called only once

File: Task2-Javascript/src/linear-unfold.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+function linearUnfold(callback, initialState) {
+    var currentState = initialState;
+	var currentArray = [];
+    while (currentState != null) {
+		currentArray.push(currentState);
+        currentState = callback(currentState);	
+    }	
+    return currentArray.slice(1);
```
Comment (ilyahumbar) [inline Task2-Javascript/src/linear-unfold.js]: Can we modify loop a bit to get rid of .slice(1); in the end of return statement?

File: Task2-Javascript/src/memoization.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+function lazyMemoEvaluation(func) {
+  var result;
+  var lazyEvaluation = func.bind.apply(func, arguments);
```
Comment (ilyahumbar) [inline Task2-Javascript/src/memoization.js]: Function should take function and must not wrap its arguments. It should work that way:
var lazyMemoSum = lazyMemoEvaluation(sum);
var result = lazyMemoSum(2, 3, 4);
var result2 =  lazyMemoSum(2, 3, 4); //function sum must not be called twice
var result3 = lazyMemoSum(2, 5);
var result3 = lazyMemoSum(2, 5); //function sum must not be called twice

File: Task2-Javascript/src/lazyMemoEvaluation.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+function lazyMemoEvaluation(func) {
```
Comment (ilyahumbar) [inline Task2-Javascript/src/lazyMemoEvaluation.js]: I'd rather put any function under require/define. 
To make it easier to access you can consider introducing 2 separate AMD modules:
1. functional extensions (one module which contains all the functions from the task)
2. any other functions required for the task (e.g. random sum etc)

File: Task2-Javascript/src/lazyMemoEvaluation.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+    var storage = [];
```
Comment (ilyahumbar) [inline Task2-Javascript/src/lazyMemoEvaluation.js]: must not be array

File: Task2-Javascript/src/lazyMemoEvaluation.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+    return function(arg) {
+        if (!(arg in storage)) {        
+            storage[arg] = func(arg);
```
Comment (ilyahumbar) [inline Task2-Javascript/src/lazyMemoEvaluation.js]: What if the argument is an object? null? undefined? what if object contains circular references? Please, add tests for all these use cases. 
e.g. 
var sum = function(obj) {
  var result = 0;
  for (var p in obj) {
    result += obj[p];
  }
  return result;
}

var  memoSum = lazyMemoEvaluation(sum);
var r = memoSum({a:1, b:2}) //3, function was run
var r1 = memoSum({a:1, b:2}) //3, function wasn't run
var r2 = memoSum({a:2, b:3}) //3, **function was run**

File: Task2-Javascript/src/lazyEvaluation.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+            if (!arg) {
+                if (!noArg) {
+                    noArg = bindedLazyEvaluation(); 
+                }
+                return false;
+            }
+            return true;
+        }
+        
+        return function(arg) {
```
Comment (ilyahumbar) [inline Task2-Javascript/src/lazyEvaluation.js]: Why this function takes any arguments? all arguments should be passed when you first call lazyEvaluation function. All arguments should be ignored.

**Comment by ilyahumbar** [comment]
Please, check all points when you are done. Check all the points in the previous messages as well

- [ ] Add additional test cases for memoization and lazy evaluation according to my comments.
- [ ] Fix implementations for memoization and lazy evaluation tasks. Ask questions if any.
- [ ] Use AMD for **_every_** function or don't use it at all. You can group your functions in different modules (e.g. Functional module etc.)
